### PR TITLE
Add CircularMean

### PR DESCRIPTION
**Pull request by rikonor**
In reference to #141

Apologies but I'm unsure of how to incorporate weights into the calculation.

### CODE REVIEW

File: stat.go [hunk 1]
```diff
@@ -372,6 +372,16 @@ func HarmonicMean(x, weights []float64) float64 {
 	return math.Exp(math.Log(W) - v)
 }
 
+// CircularMean returns the circular mean of the dataset
+func CircularMean(x []float64) float64 {
```
Comment (btracey) [inline stat.go]: It'll just be `aX += w[i] * math.Cos(x[i])` if the weights aren't nil. Look at Mean for an example how to code.

Please move this function to be in alphabetical order in the file.

File: stat_test.go [hunk 2]
```diff
@@ -333,6 +333,32 @@ func TestHarmonicMean(t *testing.T) {
 	}
 }
 
+func ExampleCircularMean() {
+	x := []float64{.0, 2 * math.Pi, 4 * math.Pi, 6 * math.Pi, 8 * math.Pi}
```
Comment (btracey) [inline stat_test.go]: This is a good example to have, but you should also have one where the mean is non zero.

Comment (btracey) [inline stat_test.go]: No need for the decimal point in front of 0

File: stat_test.go [hunk 3]
```diff
@@ -333,6 +333,32 @@ func TestHarmonicMean(t *testing.T) {
...
+}
+
+func TestCircularMean(t *testing.T) {
+	for i, test := range []struct {
+		x   []float64
+		ans float64
+	}{
+		{
+			x:   []float64{.0, 2 * math.Pi},
+			ans: .0,
```
Comment (btracey) [inline stat_test.go]: No need for the decimal points in front of 0

Comment (btracey) [inline stat_test.go]: Please also add a test with non-zero mean (and also tests with weighted data)

File: stat.go [hunk 4]
```diff
@@ -372,6 +372,16 @@ func HarmonicMean(x, weights []float64) float64 {
 	return math.Exp(math.Log(W) - v)
 }
 
+// CircularMean returns the circular mean of the dataset
```
Comment (kortschak) [inline stat.go]: Full stop please.

File: stat_test.go [hunk 5]
```diff
@@ -333,6 +333,32 @@ func TestHarmonicMean(t *testing.T) {
 	}
 }
 
+func ExampleCircularMean() {
+	x := []float64{.0, 2 * math.Pi, 4 * math.Pi, 6 * math.Pi, 8 * math.Pi}
+	cmean := CircularMean(x)
+
+	fmt.Printf("The circular mean is %.5f.\n", cmean)
+	// Output:
+	// The circular mean is -0.00000.
```
Comment (kortschak) [inline stat_test.go]: It's probably worth special-casing 0 so that -0 is not returned.

**Comment by rikonor** [comment]
Hi all,

Thanks for the quick response - I've tried to address all the issues raised.
One question - do you all agree about the special case for 0? I've tried to implement that on [stats.go#148](https://github.com/gonum/stat/pull/142/files#diff-33838b2a8ea2f8e4311d86efea5944c5R148).

File: stat.go [hunk 6]
```diff
@@ -125,6 +125,32 @@ func ChiSquare(obs, exp []float64) float64 {
 	return result
 }
 
+// CircularMean returns the circular mean of the dataset.
```
Comment (btracey) [inline stat.go]: Could you provide the formula like we do in Mean? It makes it clear for those who don't know.

File: stat.go [hunk 7]
```diff
@@ -125,6 +125,32 @@ func ChiSquare(obs, exp []float64) float64 {
...
+// If weights is nil then all of the weights are 1. If weights is not nil, then
+// len(x) must equal len(weights).
+func CircularMean(x, weights []float64) float64 {
+	if weights != nil && len(x) != len(weights) {
+		panic("stat: slice length mismatch")
+	}
+
+	var aX, aY float64
```
Comment (btracey) [inline stat.go]: Change this as @jonlawlor  suggested, and may as well move the weights check out of the loop

````
var aX, aY float64
if weights != nil {
    for _, v := range x {
        ax += math.Cos(v)
        ay += math.Sin(v)
    }
}else{
    for i, v := range x {
        aX += weights[i]  * math.Cos(v)
        aY += weights[i] * math.Sin(v)
   }
}
````

File: stat.go [hunk 8]
```diff
@@ -125,6 +125,32 @@ func ChiSquare(obs, exp []float64) float64 {
...
+			aX += weights[i] * math.Cos(x[i])
+			aY += weights[i] * math.Sin(x[i])
+			continue
+		}
+		aX += math.Cos(x[i])
+		aY += math.Sin(x[i])
+	}
+
+	res := math.Atan2(aY, aX)
+	if math.Abs(res) < 1e-14 {
```
Comment (btracey) [inline stat.go]: I'm not sure what @kortschak had in mind, but this shouldn't be based on a tolerance. Maybe

````
res := math.Atan2(aY, aX)
// correct for negative 0
if res == 0 { 
      return 0
}
````

File: stat_test.go [hunk 9]
```diff
@@ -13,6 +13,51 @@ import (
...
+		wts []float64
+		ans float64
+	}{
+		{
+			x:   []float64{0, 2 * math.Pi},
+			ans: 0,
+		},
+		{
+			x:   []float64{0, 0.5 * math.Pi},
+			ans: 0.78539816339744,
```
Comment (btracey) [inline stat_test.go]: Are these numbers from scipy or somewhere? If so, please comment.

Comment (btracey) [inline stat_test.go]: For example, https://github.com/gonum/stat/blob/master/distuv/studentst_test.go#L20

File: stat_test.go [hunk 10]
```diff
@@ -13,6 +13,51 @@ import (
 	"github.com/gonum/floats"
 )
 
+func ExampleCircularMean() {
+	x := []float64{0, 2 * math.Pi, 4 * math.Pi, 6 * math.Pi, 8 * math.Pi}
```
Comment (btracey) [inline stat_test.go]: Please include an example with numbers that aren't all zero. (maybe 0, pi/3, 2pi/3, pi or something)

File: stat.go (hunk 8 above)
Comment (rikonor) [inline stat.go]: Not sure I follow here - in the above snippet the `if` block seems redundant.

File: stat.go (hunk 7 above)
Comment (rikonor) [inline stat.go]: Moved the `if` out of the loop.

File: stat.go (hunk 6 above)
Comment (rikonor) [inline stat.go]: I added the formula, tried to rely on the formatting from the other formulas in `stats.go`.

File: stat_test.go (hunk 9 above)
Comment (jonlawlor) [inline stat_test.go]: Generally no.

**Comment by rikonor** [comment]
I hope I've addressed all the issues, is there anything I missed?

File: stat.go [hunk 11]
```diff
@@ -125,6 +125,35 @@ func ChiSquare(obs, exp []float64) float64 {
...
+// len(x) must equal len(weights).
+func CircularMean(x, weights []float64) float64 {
+	if weights != nil && len(x) != len(weights) {
+		panic("stat: slice length mismatch")
+	}
+
+	var aX, aY float64
+	if weights != nil {
+		for i := range x {
+			aX += weights[i] * math.Cos(x[i])
```
Comment (btracey) [inline stat.go]: You still need to do two-element range syntax here (as suggested before)

File: stat.go [hunk 12]
```diff
@@ -125,6 +125,35 @@ func ChiSquare(obs, exp []float64) float64 {
...
+		}
+	} else {
+		for i := range x {
+			aX += math.Cos(x[i])
+			aY += math.Sin(x[i])
+		}
+	}
+
+	res := math.Atan2(aY, aX)
+	if math.Abs(res) < 1e-14 {
```
Comment (btracey) [inline stat.go]: Don't have this tolerance (as suggested before).

File: stat_test.go (hunk 9 above)
Comment (btracey) [inline stat_test.go]: Usually we do have a comment that we got the numbers from scipy (see the link I posted above).

File: stat.go (hunk 12 above)
Comment (rikonor) [inline stat.go]: Removed.

File: stat.go (hunk 11 above)
Comment (rikonor) [inline stat.go]: Updated. Is there a difference in performance? I've noticed other places in `stat.go` follow the one-element range syntax (which is why I followed suit).

**Comment by btracey** [comment]
If you could please add the comment that the data is from scipy, then LGTM

**Comment by btracey** [comment]
And if you could also add yourself to CONTRIBUTORS and/or AUTHORS though a pull request to gonum/license

File: stat_test.go (hunk 9 above)
Comment (rikonor) [inline stat_test.go]: Verified the results via `scipy`:
```
import math
import scipy
from scipy import stats

scipy.stats.circmean(samples=[0, math.pi / 2])
```

And added a comment explaining that.

**Comment by btracey** [comment]
@jonlawlor Do you want another look?

**Comment by rikonor** [comment]
Very nice. Added a PR to `gonum/license`. Thanks guys.

**Comment by btracey** [comment]
Thanks for the contribution. Waiting for approval from @jonlawlor

File: stat.go (hunk 11 above)
Comment (jonlawlor) [inline stat.go]: There used to be a performance difference, but I don't know if the compiler has gotten smart enough in the last few Go versions to optimize.  Regardless of performance, [Effective Go](https://golang.org/doc/effective_go.html#for) says to use it:
> If you only need the second item in the range (the value), use the blank identifier, an underscore, to discard the first:
> ```
sum := 0
for _, value := range array {
    sum += value
}
```
So it is idiomatic and very slightly less complicated to use `_, v`.  If it occurs in other places in stat.go (or any gonum package) then that is a mistake and we should fix it.

File: stat_test.go [hunk 13]
```diff
@@ -13,6 +13,52 @@ import (
...
+	// The circular mean is 1.37037.
+}
+
+func TestCircularMean(t *testing.T) {
+	for i, test := range []struct {
+		x   []float64
+		wts []float64
+		ans float64
+	}{
+		// Valuess compared against scipy.
```
Comment (jonlawlor) [inline stat_test.go]: s/Valuess/Values

Comment (rikonor) [inline stat_test.go]: Oops. Fixed 👍

File: stat.go (hunk 11 above)
Comment (rikonor) [inline stat.go]: Understood. You don't see much of `x[i]` in the wild which is I why I was curious.

**Comment by jonlawlor** [comment]
LGTM
